File: pset1/credit.c
#include <stdio.h>
#include <math.h>
#include <cs50.h>

bool checksum(long number);
string which_cc(long card_no);
bool is_even(int n);

int main(void)
{
    //prompt for input
    long card_no = get_long("Number: ");

    string msg = "INVALID";
    if (checksum(card_no)) //if pass checksum
    {
        //which CC?
        msg = which_cc(card_no);
    }

    printf("%s\n", msg);
}

bool checksum(long card_no)
/*Perform Luhn's Algo - Returns True if checksum.*/
{
    int res = 0;
    int cnt = 0;
    int digit = 0;
    long divisor = 1;
    
    do
    {
        cnt++;
        divisor *= 10;
        //get each digit, starting from the end
        digit = (card_no % divisor) / (divisor / 10);

        if(is_even(cnt)) //every other digit?
        {
            digit *= 2;
            res += digit % 10;
            res += digit / 10;
        }
        else
        {
            res += digit;
        }
    }
    while (card_no/divisor > 0); 
    
    //return true if last digit of res is 0
    if (res % 10 == 0)
    {
        return true;
    }
    return false;
}

string which_cc(long card_no)
/*Return the credit card type of card_no or INVALID*/
{
    //Get nbr of digits
    double i = 0;
    long quotient;
    do
    {
        i++;
        quotient = (card_no / pow(10.0, i));
    } while (i < 18 && quotient > 0);
    
    int first_digits;
    // Amex starts with 34 or 37 | length 15 digits
    if (i == 15 )
    {
        first_digits = card_no / pow(10.0, i - 2);
        if (first_digits == 34 || first_digits ==37)
        {
            return "AMEX";
        }
    }

    // MC starts with 51-55 | length 16 digits
    else if (i == 16)
    {
        first_digits = card_no / pow(10.0, i -2);
        
        if( first_digits > 50 && first_digits < 56)
        {
            return "MASTERCARD";
        }
    }

    // VISA starts with 4 | length 13 or 16 digits
    if (i == 13 || i == 16)
    {
        first_digits = card_no / pow(10.0, i - 1);
        if (first_digits == 4)
        {
            return "VISA";
        }
    }

    //Did not match any cards.
    return "INVALID";
}

bool is_even(int n)
/*Return true if n is even*/
{
    if (n % 2 == 0)
    {
        return true;
    }
    return false;
}

File: pset6/credit/credit.py
"""credit.py
Determines whether a provided credit card number is valid according to Luhn’s
algorithm."""


def main():
    # Prompt for input
    n = input("Number: ")

    msg = "INVALID"
    if checksum(n):
        # Which CC?
        msg = which_CC(n)
    
    print(msg)


def checksum(n: str) -> bool:
    """ Perform Lunh's algo. Return True if checksum."""
    res = 0

    for i, digit in enumerate(reversed(n), start=1):
        digit = int(digit)
        # Every other digit
        if i % 2 == 0:
            digit *= 2
            res += digit % 10
            res += digit // 10
        else:
            res += digit

    # Return true if last digit of res is 0
    if res % 10 == 0:
        return True
    return False


def which_CC(n: str) -> str:
    """Return credit card type or INVALID"""
    
    # Amex starts with 34 or 37 and has 15 digits
    if len(n) == 15:
        if n[:2] == "34" or n[:2] == "37":
            return "AMEX"
    
    # MC starts with 51-55 and has 16 digits
    if len(n) == 16:
        if int(n[:2]) > 50 and int(n[:2]) < 56:
            return "MASTERCARD"
    
    # VISA starts with 4 and has 13 or 16 digits
    if len(n) == 13 or len(n) == 16:
        if n[0] == "4":
            return "VISA"
    
    # Did not match any card
    return "INVALID"


if __name__ == "__main__":
    main()